### PR TITLE
[GPU] Fix HLO op profiler: verify generated HLO modules before profiling them.

### PR DESCRIPTION
**Pull request by sergachev**
re-opening accidentally closed https://github.com/openxla/xla/pull/18391

### CODE REVIEW

**Review by dimitar-asenov — APPROVED** [review]

File: xla/service/gpu/model/hlo_op_profiler_test.cc [hunk 1]
```diff
@@ -16,6 +16,8 @@ limitations under the License.
 #include "xla/service/gpu/model/hlo_op_profiler.h"
 
 #include <gtest/gtest.h>
+#include "tsl/platform/errors.h"
```
Comment (akuegel) [inline xla/service/gpu/model/hlo_op_profiler_test.cc:19]: You also need a BUILD dependency for that header.

**Review by akuegel — APPROVED** [review]

File: xla/service/gpu/model/hlo_op_profiler_test.cc [hunk 2]
```diff
@@ -48,6 +50,12 @@ TEST_F(HloOpProfilerTest, BasicMeasurementsAreCorrect) {
             1000);
 }
 
+TEST_F(HloOpProfilerTest, UnsupportedCombinationsDoNotCrash) {
+  HloOpProfiler profiler(test_runner_);
```
Comment (akuegel) [inline xla/service/gpu/model/hlo_op_profiler_test.cc:54]: This line causes a compile error:

```
note: candidate constructor not viable: cannot bind base class object of type 'HloRunnerInterface' to derived class reference 'HloRunner &' for 1st argument
   42 |   explicit HloOpProfiler(HloRunner& runner);
```

Comment (sergachev) [reply xla/service/gpu/model/hlo_op_profiler_test.cc]: was a conflict with https://github.com/openxla/xla/commit/5ad253ae69b2a43c91bda0c1a946c0a87c1f054f

**Review by akuegel — APPROVED** [review]

**Comment by sergachev** [comment]
> The new test fails on A100 with:
> 
> ```
> hlo_op_profiler.cc:205 -  Check failed: min_duration_ > absl::ZeroDuration() (0 vs. 0) Failed to measure kernel runtime
> ```
> 
> Could you please look into this?

Does not fail for me on A100:

bazel test xla/service/gpu/model/hlo_op_profiler_test_gpu_a100

```
[ RUN      ] HloOpProfilerTest.UnsupportedCombinationsDoNotCrash
2024-10-23 10:16:49.137293: W ./xla/service/compiler.h:213] Ignoring the buffer assignment proto provided.
2024-10-23 10:16:49.149725: I xla/backends/profiler/gpu/cupti_tracer.cc:1213] CUPTI activity buffer flushed
[       OK ] HloOpProfilerTest.UnsupportedCombinationsDoNotCrash (15 ms)
```

Do you have more details?

**Comment by dimitar-asenov** [comment]
The issue is that the internal `blaze test` does not include `--config=cuda` by default, whereas it seems that in your build `--config=cuda` is implicit (maybe you have it in your `.blazerc`?).

The solution is to skip the test when cuda is not enabled. See the test just above it for how to do that.

**Comment by sergachev** [comment]
OK, done.

**Review by dimitar-asenov — APPROVED** [review]

**Pull request closed without merging**